apps/webservice/src/app/[workspaceSlug]/systems/[systemSlug]/environments/TargetFilterUniquenessIndicator.tsx

```diff
@@ -0,0 +1,206 @@
+import type { UseQueryResult } from "@tanstack/react-query";
+import type { FC } from "react";
+import type { Node } from "reactflow";
+import React, { useMemo } from "react";
+import Link from "next/link";
+import { IconAlertTriangle, IconCheck, IconLoader } from "@tabler/icons-react";
+import { useQueries } from "@tanstack/react-query";
+import LZString from "lz-string";
+
+import {
+  Tooltip,
+  TooltipContent,
+  TooltipProvider,
+  TooltipTrigger,
+} from "@ctrlplane/ui/tooltip";
+
+import { api } from "~/trpc/react";
+
+type UniqueFilterResult = {
+  isUnique: boolean;
+  overlaps: {
+    nodeId: string;
+    overlappingTargetCount: number;
+  }[];
+} | null;
+
+const useTargetFilterUniqueness = (
+  nodes: Node[],
+  workspaceId: string,
+  currentNode: Node,
+): UniqueFilterResult => {
+  const utils = api.useUtils();
+
+  const otherNodes = useMemo(
+    () => nodes.filter((node) => node.id !== currentNode.id),
+    [nodes, currentNode.id],
+  );
+
+  const overlappingQueries: UseQueryResult<number, Error>[] = useQueries({
+    queries: otherNodes.map((node) => ({
+      queryKey: [
+        "target",
+        "overlappingTargetsCount",
+        workspaceId,
+        currentNode.data.targetFilter,
+        node.data.targetFilter,
+        node.id,
+      ],
+      queryFn: () =>
+        utils.target.byWorkspaceId.list
+          .fetch({
+            workspaceId,
+            filter: {
+              type: "comparison",
+              operator: "and",
+              conditions: [
+                currentNode.data.targetFilter,
+                node.data.targetFilter,
+              ],
+            },
+          })
+          .then((result) => result.total),
+      enabled: Boolean(
+        workspaceId && currentNode.data.targetFilter && node.data.targetFilter,
+      ),
+    })),
+  });
+
+  const isLoading = overlappingQueries.some((query) => query.isLoading);
+  const isError = overlappingQueries.some((query) => query.isError);
+
+  const overlaps = useMemo(() => {
+    if (isLoading || isError) return null;
+
+    const overlappingResults = overlappingQueries
+      .map((query, index) => {
+        const count = query.data ?? 0;
+        return count > 0 && otherNodes[index]
+          ? { nodeId: otherNodes[index].id, overlappingTargetCount: count }
+          : null;
+      })
+      .filter((result) => result !== null) as {
+      nodeId: string;
+      overlappingTargetCount: number;
+    }[];
+
+    return {
+      isUnique: overlappingResults.length === 0,
+      overlaps: overlappingResults,
+    };
+  }, [isLoading, isError, overlappingQueries, otherNodes]);
+
+  return overlaps;
+};
```

Comment:
no need for memo unless you have an expect reason. Return early null outside memo
what does this unUnqiue check do. Seems wrong

Reply:
why do we need a useMemo

Reply:
return null early?

Reply:
if we need isUnique, it doesn't need to be inside the useMemo